docs/ProgrammingConcerns.md

```diff
@@ -0,0 +1,184 @@
+
+### `get_nprocs_conf`
+Use `FEXCore::CPUInfo::CalculateNumberOfCPUs` instead.
+
+### `getcwd`
+Don't use getcwd with a nullptr buffer. It will allocate memory behind our back and return a pointer that needs a free.
+
+### `strerror`
+This allocates and frees memory based on locale! Even with C local it'll attempt to free(0).
+FEX-Emu should avoid using this function and instead just return the number.
+If necessary FEX will provide its own routine for getting this string back.
+
+### `std::make_unique`
+Use `fextl::make_unique` instead.
+
+### `std::unique_ptr`
+Use `fextl::unique_ptr` instead.
+
+### `std::filesystem`
+This namespace is /highly/ likely to allocate memory behind our back.
+FEX should avoid using this API as much as possible.
+
+#### std::filesystem::path
+#### std::filesystem::path::is_relative
+Use `FHU::Filesystem::IsRelative` instead.
+
+#### std::filesystem::absolute
+Always allocates memory.
+Use `realpath` instead.
+
+#### std::filesystem::exists
+Creates a std::filesystem::path when passing in to it.
+Use `FHU::Filesystem::Exists` instead.
+
+#### std::filesystem::canonical
+Always allocates memory.
+Use `realpath` instead.
+
+#### std::filesystem::path::lexically_normal
+Use `FHU::Filesystem::LexicallyNormal` instead.
+
+#### std::filesystem::create_directory
+#### std::filesystem::create_directories
+Creates a std::filesystem::path when passing in to it.
+Use `FHU::Filesystem::CreateDirectory` and `FHU::Filesystem::CreateDirectories` instead.
+
+#### std::filesystem::path::parent_path
+Use `FHU::Filesystem::ParentPath` instead.
+
+#### std::filesystem::path::filename
+Use `FHU::Filesystem::GetFilename` instead.
+
+#### std::filesystem::copy_file
+Use `FHU::Filesystem::CopyFile` instead.
+
+### `std::fstream`
+This API always allocates memory and should be avoided.
+Use a combination of open and fextl::string APIs instead of fstream.
+
+### `std::fwrite`
+This API allocates a buffer in the background for buffering output.
+Use raw `write` instead.
+
+### `std::string`
+Use `fextl::string` instead.
+
+#### std::to_string
+Use `fextl::fmt::format` instead.
+
+### `std::stol`
+### `std::stoul`
+### `std::stoll`
+### `std::stoull`
+These all consume a `std::string` as their first argument. Use the equivalent functions that don't use `std::string`
+- `std::strtol`
+- `std::strtoul`
+- `std::strtoll`
+- `std::strtoull`
+
+### `fmt::`
+### `fmt::format`
+Use `fextl::fmt::` instead
+
+### APIs that FEX doesn't have a replacement for
+Don't use any of these APIs in FEXLoader/FEXInterpreter. Shoutout to
+[this](https://stackoverflow.com/questions/43056338/standard-library-facilities-which-allocate-but-dont-use-an-allocator) StackOverflow post for this
+huge list.
+
+#### `std::any`
+#### `std::function` and lambdas
+One must take additional considerations when using these to ensure that they don't allocate memory. These don't have any way to replace which
```

Comment:
Lambdas never heap-allocate memory as far as I'm aware, even with large capture lists. So the use of `these` is a bit misleading.

What's dangerous (and worth highlighting hence) is passing an innocent-looking lambda as a `std::function` parameter, though.

Reply:
Lambdas can allocate memory. I had to change this lambda because it was allocating memory: https://github.com/FEX-Emu/FEX/blob/main/External/FEXCore/Source/Interface/Config/Config.cpp#L725
Was allocating memory to store `EnvMapType EnvMap` from the capture list.

Reply:
Are you sure this was actually from the lambda itself and not e.g. from accidentally copying the EnvMapType into a non-fextl container when capturing? `[&EnvMap]() {...}` really is just syntactic sugar for creating a stack-allocated class with the capture list as members, so I don't see how this possible could lead heap-allocate.

Reply:
Checking https://github.com/FEX-Emu/FEX/commit/1f9458a3c3c61de68b886e1e089db2c9f5a3835c, it looks like the lambda was assigned to an `std::function` before. The lambda itself was still stack-allocated, but `std::function` would copy the large lambda to heap-allocated storage.

Reply:
ah, that would make sense.